Automattic/wp-calypso
# State: Refactor preferences to split local and remote values

Related: #7478 (reverted in #7556)

This pull request seeks to refactor the preferences state to better distinguish local and remote preferences. In Calypso, some local preferences exist which are not intended to be saved to the REST API user settings endpoint. For example, `mediaModalGalleryInstructionsDismissedForSession` is a preference which controls the visibility of the media gallery help dialog, but is only saved for the duration of the current page session.

The issue with a combined preference state is that it is difficult to determine whether preferences have been received yet from the REST API. It's not enough to rely on `fetching` state, since this check can occur before the query component for preferences has been mounted. In #7478, this is an issue [in the sites controller](https://github.com/Automattic/wp-calypso/blob/72d15ff8fb7c727806f30937b37f5e49eea64795/client/my-sites/controller.js#L176) which needs to be able to guard against cases where a request is in progress (where updating the preference would clobber existing values) while still allowing the preference to be set if the user does not currently have a recent sites preference value.

__Testing instructions:__

Verify that there are no regressions in Redux-based preferences:
- [x] Guided tours history
- [x] First view
- [x] Media gallery dismissal
- [x] Editor tab

Ensure that Mocha tests pass:

```
npm run test-client
```

/cc @gwwar @artpi 

Test live: https://calypso.live/?branch=update/preferences-received
